### PR TITLE
feat(xml): The more robust and clearer multi-pass parser

### PR DESCRIPTION
**Pull request by huieric**
The xml parser was refactored into a three-pass parser:
1. Read xml file with `xmlquery`, construct a recursively self-described tree structure defined in `xml.proto` and put it into memory.
2. Preprocess the tree structure. In this phase the parser will do some necessary jobs before generating a 2-dimensional sheet, like `correctType` which make the types of attributes in the nodes meet the requirements of `protogen`.
3. Transform the recursive tree structure into a 2-dimensional sheet, which can be further processed into `protoconf`.

### CODE REVIEW

**Review by wenchy — CHANGES_REQUESTED** [review]

File: test/dev/proto/feature_toggle_conf.proto [hunk 1]
```diff
@@ -23,14 +23,13 @@ message ToggleCfg {
       protoconf.FeatureToggleMacroType id = 1 [(tableau.field) = {name:"Id"}];
       uint32 world_id = 2 [(tableau.field) = {name:"WorldID"}];
       uint32 zone_id = 3 [(tableau.field) = {name:"ZoneID"}];
-      google.protobuf.Timestamp open_time = 4 [(tableau.field) = {name:"OpenTime"}];
-      google.protobuf.Timestamp close_time = 5 [(tableau.field) = {name:"CloseTime"}];
-      int64 sys_key_1 = 6 [(tableau.field) = {name:"SysKey1"}];
-      int64 sys_key_2 = 7 [(tableau.field) = {name:"SysKey2"}];
+      google.protobuf.Timestamp open_time = 4 [(tableau.field) = {name:"OpenTime" prop:{default:"2022-01-23 15:40:00"}}];
+      google.protobuf.Timestamp close_time = 5 [(tableau.field) = {name:"CloseTime" prop:{default:"2022-01-23 15:40:00"}}];
```
[the remark this answers is not in the document]
Comment (huieric) [reply test/dev/proto/feature_toggle_conf.proto:27]: done

File: proto/tableau/protobuf/xml.proto [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+// Combine list and map to get an ordered map
+message OrderedAttrMap {
+  repeated Attr list = 1;
+  map<string, int32> map = 2;
+}
+
+message Attr {
+  string name = 1;
+  string value = 2;
+}
```
[the remark this answers is not in the document]
Comment (huieric) [reply proto/tableau/protobuf/xml.proto:59]: done

File: proto/tableau/protobuf/xml.proto [hunk 3]
```diff
@@ -0,0 +1,59 @@
+syntax = "proto3";
+package tableau;
+
+option go_package = "github.com/tableauio/tableau/proto/tableaupb";
+
+message XMLMeta {
```
[the remark this answers is not in the document]
Comment (huieric) [reply proto/tableau/protobuf/xml.proto:6]: done

File: proto/tableau/protobuf/xml.proto [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  // NodeMap maps path to a list of nodes matched the specified path in the tree
+  map<string, Node> meta_node_map = 3;
+  map<string, NodeList> data_node_map = 4;
+
+  message NodeList {
+    repeated Node nodes = 1;
+  }
+}
+
+message Node {
```
[the remark this answers is not in the document]
Comment (huieric) [reply proto/tableau/protobuf/xml.proto:28]: done

File: internal/importer/xml.go [hunk 5]
```diff
@@ -321,248 +372,213 @@ func secondParseSheet(root *xmlquery.NodeNavigator, metaSheet *xlsxgen.MetaSheet
...
+}
+
+func isFirstChild(node *tableaupb.Node) bool {
+	if node.Parent == nil {
+		return false
+	}
+	return node.Parent.ChildList[0] == node
+}
+
+func correctType(xmlSheet *tableaupb.XMLSheet, curr *tableaupb.Node, oriType string) (t string) {
```
[the remark this answers is not in the document]
Comment (huieric) [reply internal/importer/xml.go]: done

**Review by wenchy — APPROVED** [review]

**Pull request merged**